src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
 
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+        break;
+    }
```

Comment:
Consider adding a default case in the switch statement to handle any unexpected action types. This would improve error handling and provide a clearer message when an action with an unsupported type is used.